### DISCUSSION TITLE
Failure for Custom UploadFile Classes to Validate

### DISCUSSION THREAD

**Discussion by ccherrad** [fastapi/fastapi#11905] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
class ValidFile(UploadFile):
    """
    Wrapper for FastAPI UploadFile.
    """
    allowed_extensions = []

    @classmethod
    def validate(cls: Type["UploadFile"], v: Any) -> Any:
        if cls.allowed_extensions and not all(
            ext in SUPPORTED_EXTENSIONS for ext in cls.allowed_extensions
        ):
            raise HTTPException(
                status.HTTP_415_UNSUPPORTED_MEDIA_TYPE,
                detail="the platform doesn't support that file type.",
            )
        contents = v.file.read()
        v.extensions = from_buffer(contents)
        v.is_fitzable = bool(set(v.extensions) & set(FITZ_SUPPORTED_FILETYPES))
        v.file.seek(0)
        if not (set(v.extensions) & set(cls.allowed_extensions)):
            raise HTTPException(
                status.HTTP_415_UNSUPPORTED_MEDIA_TYPE,
                detail="file type not supported.",
            )
        return super(ValidFile, cls).validate(v)

class ValidZipFile(ValidFile):
    """
    Wrapper for FastAPI Zip UploadFile.
    """
    allowed_extensions = ["zip"]

@router.post(
    "/",
    tags=["Workflow"],
    status_code=status.HTTP_201_CREATED,
)
def start_workflow(
    document: ValidZipFile,
):
    # worker.send_task("get_information", (documents,))

    return document
```

### Description

I encountered a regression in the UploadFile validate function when upgrading FastAPI to 0.111.1, and Pydantic to version 2.
I have created a custom class ValidFile inheriting from UploadFile to perform additional validation on uploaded files. I then subclassed ValidFile to create ValidZipFile with specific allowed extensions.

the same thing woks fine with these versions: 
fastapi==0.109.2
pydantic==1.10.14
python==3.8.19

### Operating System

Linux

### FastAPI Version

0.111.1

### Pydantic Version

2.8.2

### Python Version

3.9.19

### Additional Context

I noticed that the only things that was changed between the two version is the function: request_body_to_args

**Comment by Kludex** [comment]
If you give me a code that I can run, and present me the actual and expected behavior, then I can try to help.

**Reply by ccherrad** [reply]
yes sure, do this is an example: https://gist.github.com/ccherrad/d5345edeeea20be7efc30da0b949f2d3
thank you so much for your reply.

**Reply by Kludex** [reply, marked as answer]
This is not because of FastAPI bump, but because you bumped Pydantic on the same dependencies bump.

Here's how you'd achieve the same without many changes:

```py
from __future__ import annotations

from typing_extensions import Any

from fastapi import FastAPI, HTTPException, UploadFile, status, APIRouter

import magic


class ValidFile(UploadFile):
    """
    Wrapper for FastAPI UploadFile.
    """

    allowed_extensions: list[str] = []

    @classmethod
    def _validate(cls, __input_value: Any, _: Any) -> "UploadFile":
        v: UploadFile = super()._validate(__input_value, _)
        contents = v.file.read()

        # This actually returns a single value...
        extension = magic.from_buffer(contents)
        print(extension)

        if not extension in cls.allowed_extensions:
            raise HTTPException(
                status.HTTP_415_UNSUPPORTED_MEDIA_TYPE,
                detail="file type not supported.",
            )
        v.file.seek(0)

        return v


class ValidPythonFile(ValidFile):
    """
    Wrapper for FastAPI PDF UploadFile.
    """

    allowed_extensions = ["Python script, ASCII text executable"]


router = APIRouter(prefix="/documents")


@router.post(
    "/",
    tags=["Documents"],
    status_code=status.HTTP_201_CREATED,
)
def post_document(document: ValidPythonFile):
    return "we are fine it is a PDF file"


app = FastAPI(title="documents")

app.include_router(router)
```

This is not the way I would recommend tho. I think you need to create dependencies, and then pass the `UploadFile` to them, and do the validation, instead of the above. You can even have a factory of `Depends`cies.

Also, try to reproduce your own code before asking help to someone else. Your code was not running.

**Reply by ccherrad** [reply]
sorry, I created a Gist from my code and used Magic, but I have my own magic wrappers that return List instead of str.
thank you I think I understand now. I should use _validate instead of validate when upgrading to pydantic 2.
